Ряд полезностей

ili opened the pull request:
* Поддержка Фреймворка 4.5
* Поддержка `async`
* Фреймворк 4.5+ использует для вызовов `HttpClient`
* Поддержка запроса о состоянии платежа
* Тестовое приложение под `netcoreapp2.0`
* По мелочи документации
* Создание платежа с чеком (под ФЗ-54)
* Не обрабатывался ответ от сервера, теперь ошибки обрабатываются и пакуются в `YandexCheckoutException`

ili committed:
.editorconfig

ili committed:
documentation

receipt for federal law 54

ili committed:
version

ili committed:
async extensions

net 452

ili committed:
.Net 4 sample

ili committed:
Using HttpClient for net45+

introducing YandexCheckoutException

ili committed:
YandexCheckoutException now used only for error responce from server

connection errors would not be wrapped

ili committed:
netcoreapp2.0 test application

bzaar commented:
Илья, спасибо за PR! Изменения полезные, смущает только разбухание кода от `#if NET40 ... #else...` и оттого, что теперь каждый метод в sync/async вариантах (через copy-paste). Думаю, этого можно избежать. Класс Task есть в .NET 4.0, класс HttpClient - тоже, в виде отдельного [Nuget-пакета](https://www.nuget.org/packages/Microsoft.Net.Http/). Также есть пакет [поддержки async/await](https://stackoverflow.com/a/19421907/1291717) в .NET 4.0.

Есть желание эту проблему исправить? Или я сам посмотрю при случае.

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -115,53 +115,54 @@ public Task<Payment> CaptureAsync(Payment payment, string idempotenceKey = null)
+        }
 
-                return JsonConvert.DeserializeObject<T>(responceData);
+        private static T ProcessResponce<T>(HttpResponseMessage response, string responceData)
```
Опечатка: responce => response. В других местах тоже.

ili replied:
Исправил

bzaar commented:
`wwwroot\lib` needs to be removed and .gitignore'd.

ili commented:
Да, что-то я забыл о наличии BCL для NET 4. Это нужно будет поправить.  А вот sync\async это боюсь неизбежный копипаст, стремился его минимизировать, вынеся в асинхронщину только запросы к апи... но как бы тренд, в наше время без асинка никуда :(

в общем-то в любом случае я буду доделывать ещё методы возврата, так что пока помечу этот реквест как WIP

Есть пара рабочих вопросов:
* методы получения инфо о платеже и подтверждения платежа требуют только идентификатор платежа, так что я бы предложил изменить интерфейс соответствующих вызовов, на текущий момент он избыточен, и как следствие усложняет интеграцию (нужно либо заново строить платёж, либо где-то его хранить. Я предлагаю минимизировать контракт до:
```cs
Capture(string id, string idempotenceKey)
QueryPayment(string id, string idempotenceKey)
```
* использование `Guid` в качестве `id` - в документации нигде не сказано, что это валидный `GUID` (ну либо я не нашел), т.е. вполне может быть что "завтра" это превратится в Base64 строку какого-нить хэша, или, скажем к нему добавится идентификатор магазина или ещё что. Предлагаю сменить на `string`
* язык XML комментариев - где-то английский где-то русский, нужен один
* нейминг - snake_case - не используется в дотнете, я бы предложил перейти на CamelCase с развешиванием атрибутов на свойства

bzaar commented:
Sync методы можно реализовать вызовом соответствующего async метода через
await. Тогда копипаст не нужен.

On Thu, 11 Oct 2018, 07:54 Ilya Chudin, <notifications@github.com> wrote:

> Да, что-то я забыл о наличии BCL для NET 4. Это нужно будет поправить. А
> вот sync\async это боюсь неизбежный копипаст, стремился его минимизировать,
> вынеся в асинхронщину только запросы к апи... но как бы тренд, в наше время
> без асинка никуда :(
>
> в общем-то в любом случае я буду доделывать ещё методы возврата, так что
> пока помечу этот реквест как WIP
>
> Есть пара рабочих вопросов:
>
>    - методы получения инфо о платеже и подтверждения платежа требуют
>    только идентификатор платежа, так что я бы предложил изменить интерфейс
>    соответствующих вызовов, на текущий момент он избыточен, и как следствие
>    усложняет интеграцию (нужно либо заново строить платёж, либо где-то его
>    хранить
>    - язык XML комментариев - где-то английский где-то русский, нужен один
>    - нейминг - snake case - не используется в дотнете, я бы предложил
>    перейти на CamelCase с развешиванием атрибутов на свойства
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#issuecomment-428841967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD15-kF0d_rcSfGCkpuTJdak5B7SP9w1ks5ujusWgaJpZM4XTw_n>
> .
>

bzaar commented:
Про методы получения инфо о платеже и подтверждения платежа согласен.
Совместимость будет сломана, но ничего.

On Thu, 11 Oct 2018, 07:54 Ilya Chudin, <notifications@github.com> wrote:

> Да, что-то я забыл о наличии BCL для NET 4. Это нужно будет поправить. А
> вот sync\async это боюсь неизбежный копипаст, стремился его минимизировать,
> вынеся в асинхронщину только запросы к апи... но как бы тренд, в наше время
> без асинка никуда :(
>
> в общем-то в любом случае я буду доделывать ещё методы возврата, так что
> пока помечу этот реквест как WIP
>
> Есть пара рабочих вопросов:
>
>    - методы получения инфо о платеже и подтверждения платежа требуют
>    только идентификатор платежа, так что я бы предложил изменить интерфейс
>    соответствующих вызовов, на текущий момент он избыточен, и как следствие
>    усложняет интеграцию (нужно либо заново строить платёж, либо где-то его
>    хранить
>    - язык XML комментариев - где-то английский где-то русский, нужен один
>    - нейминг - snake case - не используется в дотнете, я бы предложил
>    перейти на CamelCase с развешиванием атрибутов на свойства
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#issuecomment-428841967>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD15-kF0d_rcSfGCkpuTJdak5B7SP9w1ks5ujusWgaJpZM4XTw_n>
> .
>

ili commented:
Вообще вопрос в том нужна ли поддержка 4 Фреймворка. В принципе если цеплять BCL то можно выпилить поддержку синхронных методов. 
Вообще же не желательно вместо прокидывать асинк методы через Result не лучшее решение, т.к. есть шанс получить дедлок при вызове в потоке UI... т.е. в целях стабильности я бы даже, пожалуй, все синхронные методы бы прогнал через HttpWebRequest...
Т.е. размышляя я бы принес копипасту в жертву надежности

bzaar commented:
Про русские/английские XML комментарии - можно использовать

        /// <summary>English summary</summary>
        /// <summary xml:lang='ru'>Описание по-русски</summary>

([отсюда](https://stackoverflow.com/q/6221140/1291717)). Есть даже инструменты ([Surviveplus.net](http://www.surviveplus.net/en/archives/39)) для генерации XML comment files на разных языках. 

Сам Яндекс поддерживает документацию на двух языках.

bzaar commented:
Поддержка 4.0 нужна, например, мне: есть у меня один довольно старый и
большой сайт. Возможно, я не один такой. Так что пока поддержка 4.0 в
клиенте не вызывает особых трудностей, я бы ее оставил.

bzaar commented:
> В принципе если цеплять BCL то можно выпилить поддержку синхронных
методов.

"Выпилить" = удалить? Наоборот BCL позволяет использовать async/await в
.NET 4.0.

Насчет дедлока - просто не надо вызывать синхронные методы, если у вас
приложение асинхронное. Синхронные они по определению блокируют вызывающий
поток. Так что не вижу проблемы в сосуществовании синхронных и асинхронных
методов. Необходимость же есть в обоих.

On Thu, 11 Oct 2018, 11:34 Ilya Chudin, <notifications@github.com> wrote:

> Вообще вопрос в том нужна ли поддержка 4 Фреймворка. В принципе если
> цеплять BCL то можно выпилить поддержку синхронных методов.
> Вообще же не желательно вместо прокидывать асинк методы через Result не
> лучшее решение, т.к. есть шанс получить дедлок при вызове в потоке UI...
> т.е. в целях стабильности я бы даже, пожалуй, все синхронные методы бы
> прогнал через HttpWebRequest...
> Т.е. размышляя я бы принес копипасту в жертву надежности
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#issuecomment-428906404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD15-u3LGMMRCgR9DzM9_h6SD2J3WK4Uks5ujx68gaJpZM4XTw_n>
> .
>

bzaar commented:
Про snake_case - не против навесить атрибуты.

bzaar commented:
Про Guid -> string согласен.

ili commented:
>Насчет дедлока - просто не надо вызывать синхронные методы, если у вас
приложение асинхронное. Синхронные они по определению блокируют вызывающий
поток. Так что не вижу проблемы в сосуществовании синхронных и асинхронных
методов. Необходимость же есть в обоих.

Нет, я вот про такой вызов асинхронного метода из синхронного:
```
public int SomeFunc()
{
    var res  = CallAsync().Result; 
}
```

вот он может привести к очень неприятным последствиям (тупо зависнет приложения, сам сталкивался). Вот я здраво подумав и пришел к выводу что для синхронных версий методов лучше использовать `HttpWebRequest` вместо `HttpСlient` - и бог бы с ней с небольшой копипастой, надёжней будет (вот сейчас я правда и пользую `.Result` носовести моей это не дает покоя, так что думаю лучше переделать)

ili committed:
Использование CamelCase в именах

ili commented:
> wwwroot\lib needs to be removed and .gitignore'd.

в общем тут некоторый оверхэд - в дефолтный шаблон приложения не включен ни bower, ни npm... т.е. для боевого приложения я бы конечно запарился с подключением либ и всё такое прочее, но для тестового на 5 страниц... можно, конечно поправить загрузку и сделать её безусловной всегда из CDN... но опять таки... времени жалко...

ili committed:
Изменены контракты получения и подтверждения платежа

Теперь передается не полная информация о платеже а только идентификатор

bzaar reviewed (commented):
>     var res  = CallAsync().Result; 

Есть несколько решений этой проблемы:

 * [ConfigureAwait(false)](https://medium.com/bynder-tech/c-why-you-should-use-configureawait-false-in-your-library-code-d7837dce3d7f)
 * [AsyncHelper.RedirectToThreadPool()](https://ru.stackoverflow.com/a/681466/194429)

ili committed:
Задел под поддержку BCL для net40

ili committed:
Поправил синхронный запрос без тела

ili committed:
Уменьшил дублирование кода в примере

ili commented:
В последних изменениях:
* изменил контракты ф-й получения инфо о платеже и подтверждения платежа
* настроил серилизатор на использования snake_case, а все модели перевёл в CamelCase (аналогично с `enum`, обвешал их соответствующими атрибутами)
* id платежа теперь строка
* а вот с async для NET40 возникли проблемы - оно банально не собирается при подключении BCL, честно признаться сейчас нет времени с этим разбираться, так что всё необходимое я пихнул под условную компиляцию, и для включения поддержки async достаточно в файле проекта изменить `<Net40AsyncSupport>NO</Net40AsyncSupport>` на `<Net40AsyncSupport>YES</Net40AsyncSupport>`
* так же синхронные и асинхронные методы в потрахах пользуют разные механизмы - `HttpClient` & `HttpWebRequest`, ИМХО надежность важнее дублирования кода в данном случае

ili commented:
>Есть несколько решений этой проблемы:

И все они приводили у меня к зависанию программы. Я тут просто с несколько месяцев назад перетаскивал текущий проект с синхронных на асинхронные вызовы... и местами дабы "задешманить" пробовал эти варианты... зависало

ili committed:
Исправил опечатку

ili committed:
Ворнинги как ошибки

ili commented:
Ну и не собирается оно с BCL...

bzaar commented:
Спасибо за оперативность!  А что говорит при попытке собрать с BCL?

On Fri, 12 Oct 2018, 08:43 Ilya Chudin, <notifications@github.com> wrote:

> Ну и не собирается оно с BCL...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#issuecomment-429235730>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD15-jWNfR53CFOO8uaR1ePraaJDJ3_fks5ukEg0gaJpZM4XTw_n>
> .
>

bzaar committed:
Устранены некоторые опечатки и уменьшено количество копипаста.

bzaar committed:
Опечатки в TaxSystem

bzaar committed:
Receipt перемещен в NewPayment. PaymentWithReceipt удален.

bzaar committed:
Опечатка: плетельщик

bzaar committed:
<Net40AsyncSupport>YES</Net40AsyncSupport> у меня работает вроде.

bzaar committed:
async заработал под .NET 4.0

ili commented:
> Спасибо за оперативность!  А что говорит при попытке собрать с BCL?
 
что не может найти метод или расширение GetAwaiter<HttpResponceMessage>

ili committed:
Merge pull request #1 from bzaar/feature/async

Feature/async

ili committed:
Версия 0.0.6

ili committed:
Выключил поддержку асинка для .Net 4.0

ili commented:
Пока вроде как всё.
Далее в планах возвраты, но это на горизонте пары недель

ili committed:
Добавлен метод овормления возврата

Так же добавлена валидация параметров конструктора Client

ili committed:
Более полная поддержка API

* Отмена платежа
* Передача чека в структуру возврата (54-ФЗ)
* Подтверждение платежа (capture) с возможностью изменения суммы

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -1,37 +1,225 @@
+        /// <param name="refound">Refound data</param>
+        /// <param name="idempotenceKey">Idempotence key, use <value>null</value> to generate new one</param>
+        /// <returns><see cref="NewRefound"/></returns>
+        public Refound Refound(NewRefound refound, string idempotenceKey = null)
```
В слове Refund не должно быть буквы 'o'.

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -88,6 +88,16 @@ public Payment CreatePayment(NewPayment payment, string idempotenceKey = null)
+        
+        /// <summary>
+        /// Payment capture, can be used to change payment amount.
+        /// If you do not need to make any changes in paymnet use <see cref="Capture(string,string)"/>
```
Опечатка: paymnet

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -88,6 +88,16 @@ public Payment CreatePayment(NewPayment payment, string idempotenceKey = null)
+        /// <param name="idempotenceKey">Idempotence key, use <value>null</value> to generate new one</param>
+        /// <returns><see cref="Payment"/></returns>
+        public Payment Capture(Payment payment, string idempotenceKey = null)
+            => Query<Payment>("POST", payment, _apiUrl + payment.Id + "/capture", idempotenceKey);
```
Для лаконичности я бы просто добавил к Capture(string,string) необязательный параметр типа Amount. https://kassa.yandex.ru/docs/checkout-api/#podtwerzhdenie-platezha

ili replied:
При 54-ФЗ нужно же ещё состав чека менять, так что тут одним Amount по идее не отделаешься

ili committed:
Исправлены возвраты

* Возвраты делались по неверному адресу
* Обработка ошибок теперь проверяет поддерживаемые коды и тип контента (при обращении по неверному адресу можем получить 405, где в ответе будет не JSON)
* Client.ParseMessage теперь проверяет только начало контент тайпа

ili committed:
Электропочта для чека

ili committed:
Исправлены опечатки

bzaar commented:
По ссылке в моем комментарии есть пример запроса и там только amount. Где
написано , что можно менять состав чека?

On Fri, 19 Oct 2018, 08:50 Ilya Chudin, <notifications@github.com> wrote:

> *@ili* commented on this pull request.
> ------------------------------
>
> In Yandex.Checkout.V3/Client.cs
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#discussion_r226559660>
> :
>
> > @@ -88,6 +88,16 @@ public Payment CreatePayment(NewPayment payment, string idempotenceKey = null)
>          /// <returns><see cref="Payment"/></returns>
>          public Payment Capture(string id, string idempotenceKey = null)
>              => Query<Payment>("POST", null, _apiUrl + id + "/capture", idempotenceKey);
> +
> +        /// <summary>
> +        /// Payment capture, can be used to change payment amount.
> +        /// If you do not need to make any changes in paymnet use <see cref="Capture(string,string)"/>
> +        /// </summary>
> +        /// <param name="payment">New payment data</param>
> +        /// <param name="idempotenceKey">Idempotence key, use <value>null</value> to generate new one</param>
> +        /// <returns><see cref="Payment"/></returns>
> +        public Payment Capture(Payment payment, string idempotenceKey = null)
> +            => Query<Payment>("POST", payment, _apiUrl + payment.Id + "/capture", idempotenceKey);
>
> При 54-ФЗ нужно же ещё состав чека менять, так что тут одним Amount по
> идее не отделаешься
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/morpher-ru/Yandex.Checkout.V3/pull/1#discussion_r226559660>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD15-tRxDN0q7g-1LfrfZD8TZYIjJ8RMks5umYRAgaJpZM4XTw_n>
> .
>

ili committed:
Более полная поддержка API

* Типы данных запросов и ответов теперь более соответствуют API (частичная избыточность)
* Добавлен метод получения информации о возврате
* К методам Capture и Cancel добавлен постфикс Payment дабы не возникало разночтений

ili committed:
Обновлена версия

ili commented:
Тут: https://kassa.yandex.ru/docs/checkout-api/#podtwerzhdenie-platezha

ili commented:
Вообще сейчас в модели есть избыточность, по хорошему её нужно приводить к однозначному виду,  при чём с инициализацией через конструктор, но это прям закат солнца вручную (

bzaar committed:
Remove static: HttpClient is not thread-safe.

bzaar committed:
Remove extra brackets.

bzaar committed:
Use enum HttpStatusCode instead of int consistently.

bzaar committed:
Format code.

bzaar committed:
Extract string constant "application/json". Do not set request.Content when body == null.

bzaar committed:
Rename CreateAsyncRequest to CreateRequest. The request being created is HttpRequestMessage which can be sent in either manner, sync or async.

bzaar committed:
BUGFIX: Remove double slashes in //refund.

ili committed:
Ключ идемпотентности только там где нужно

ili committed:
Версия

bzaar committed:
Merge branch 'master' of https://github.com/ili/Yandex.Checkout.V3

# Conflicts:
#	Yandex.Checkout.V3/Client.cs

bzaar committed:
Remove extra backslashes.

bzaar committed:
Use HttpClient.BaseAddress.

bzaar committed:
Remove NewGuid - idempotenceKey is never null in this case.

bzaar committed:
Implement IDisposable on Client.

bzaar committed:
Revert "Use HttpClient.BaseAddress."

This reverts commit 0bafc8fcba09127a5645e79c3a3c980b54f31a1b.

bzaar committed:
Покажем хороший пример - сделаем Client static.

bzaar committed:
Remove unneeded usings.

ili committed:
Merge pull request #2 from bzaar/master

 BUGFIX: Remove double slashes in //refund.

ili committed:
Добавлен конструктор, принимающий HttpClient

Если HttpClient не передан, будет создан локальный экземпляр, который будет задиспоужен при диспоузе клиента

ili committed:
Признаки предмета и способа расчета для позиции чека

ili committed:
Удален packages.config

ili committed:
Исправлено создание HttpClient

Almeonamy commented:
Добрый день. В этом PR реализовано много хороших фич. В том числе поддержка онлайн-касс. Когда ожидается слияние этого с master и обновление nuget? Хотелось бы увидеть это в близком будущем, т.к. возникла реальная необходимость отправлять данные по онлайн-кассе. После этого я бы, возможно, тоже смог присоединиться к доработке недостающих фич.

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -1,37 +1,282 @@
+
+        /// <inheritdoc cref="CreatePaymentAsync(Yandex.Checkout.V3.NewPayment,System.Threading.CancellationToken,string)"/>
+        public Task<Payment> CreatePaymentAsync(NewPayment payment, string idempotenceKey = null)
+            => CreatePaymentAsync(payment, CancellationToken.None, idempotenceKey);
```
Объединить эти два метода в один с помощью необязательного параметра CancellationToken token = default. См. также https://stackoverflow.com/a/32634213/1291717

bzaar commented:
@Almeonamy, да, PR действительно полезный и нужный. Что мне хотелось бы доработать перед релизом - так это сделать два отдельных класса Client и AsyncClient, т.к. в 99% случаев нужен либо тот, либо другой. Преимущества этого подхода:

* Для пользователей: нужно один раз решить, какой клиент нужен. При текущем подходе пользователь каждый раз ставя точку после "client" будет думать, какой же метод мне нужен? И есть вероятность ошибиться.
* Для разработчиков (т.е. нас): упростится реализация, исчезнут #if !SYNCONLY из кода. По сути будет только один if - в csproj: 

```
  <ItemGroup Condition=" '$(TargetFramework)' != 'net40'">
      <Compile Include="AsyncClient.cs" />
  </ItemGroup>
```

Кто может и хочет это реализовать?

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -1,37 +1,282 @@
+                if (_httpClient == null)
+                {
+                    _httpClient = new HttpClient();
+                    _disposeHttpClient = true;
```
Этот код не потокобезопасный. Теоретически может создаться несколько HttpClient-ов. Я бы предложил сделать оба поля readonly и инициализировать в конструкторе.

bzaar commented on Yandex.Checkout.V3/Client.cs:
```diff
@@ -1,37 +1,282 @@
+        /// <param name="apiUrl">API URL</param>
+        /// <param name="userAgent">Agent name</param>
+        public Client(
+            HttpClient httpClient,
```
Думаю, лучше сделать параметр httpClient необязательным со значение по умолчанию null (Client сам создает HttpClient и делает ему Dispose).
